tooling/profiler/src/cli/gates_flamegraph_cmd.rs

```diff
@@ -75,15 +75,20 @@ fn run_with_provider<Provider: GatesProvider, Generator: FlamegraphGenerator>(
 
     // We can have repeated names if there are functions with the same name in different
     // modules or functions that use generics.
-    let mut seen_names: HashMap<&str, usize> = HashMap::default();
+    let mut seen_names: HashMap<String, usize> = HashMap::default();
     for (func_idx, (func_gates, bytecode)) in
         backend_gates_response.functions.into_iter().zip(bytecode.functions).enumerate()
     {
-        let func_name = function_names[func_idx].as_str();
-        let count = seen_names.entry(func_name).and_modify(|c| *c += 1).or_insert(0);
-
-        let func_name =
-            if *count == 0 { func_name.to_owned() } else { format!("{}_{}", func_name, count) };
+        let original_name = function_names[func_idx].as_str();
+        let mut function_name = original_name.to_owned();
+
+        let mut counter = seen_names.get(original_name).copied().unwrap_or(0);
+        // Ensure uniqueness by checking existing names
+        while seen_names.contains_key(&function_name) {
+            counter += 1;
+            function_name = format!("{}_{}", original_name, counter);
+        }
```

Comment:
We could also avoid the loop entirely if we separate with a suffix normal names can't use. E.g. `write_to_file.1.svg`

Reply:
I also looked making the names unique by just associating the brillig fn index with the name automatically, which also would get rid of the loop. So in the situation you laid our earlier we would end up with `write_to_array_0_brillig`, `write_to_array_1_brillig`, `write_to_array_1_2_brillig`.

Reply:
I think I may switch to that to avoid the loop. We can then refactor to switch to placing the function id in path's suffix later.